pkg/nfspv/nfsvolume.go

```diff
@@ -0,0 +1,340 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/strategicpatch"
+	"k8s.io/client-go/kubernetes"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+)
+
+const (
+	OpenEBSCASLabelKey      = "openebs.io/cas-type"
+	OpenEBSNFSCASLabelValue = "nfs-kernel"
+)
+
+var (
+	supportedDataTypes = []collectorinterface.DataType{collectorinterface.JSONDataType}
+)
+
+// nfsVolume will implement necessary methods
+// required to satisfy collector interface
+type nfsVolume struct {
+	clientset kubernetes.Interface
+
+	pvcLister corev1listers.PersistentVolumeClaimLister
+
+	pvLister           corev1listers.PersistentVolumeLister
+	pvObj              *corev1.PersistentVolume
+	nfsServerNamespace string
+	annotationPrefix   string
+	// dataType represents the type of the data that server
+	// can understand. As of now JSON is supported
+	dataType collectorinterface.DataType
+}
+
+func NewNFSVolume(
+	clientset kubernetes.Interface,
+	pvcLister corev1listers.PersistentVolumeClaimLister,
+	pvLister corev1listers.PersistentVolumeLister,
+	pvObj *corev1.PersistentVolume,
+	nfsServerNamespace string,
+	dataType collectorinterface.DataType) collectorinterface.VolumeEventCollector {
+	return &nfsVolume{
+		clientset:          clientset,
+		pvcLister:          pvcLister,
+		pvLister:           pvLister,
+		pvObj:              pvObj,
+		nfsServerNamespace: nfsServerNamespace,
+		annotationPrefix:   "nfs.",
+		dataType:           dataType,
+	}
+}
+
+// CollectCreateEvents returns the serialized data(JSON/YAML) with
+// volume creation timestamps. Serialized data must understood by
+// server.
+// NOTE: There can be a case where exporter is unable to send events
+//		 to server mean time if user deletes volume then resources
+//		 will have deletion timestamp. To avoid sending deletion
+//		 timestamp for create event we are mutating deletion timestamp
+//		 fields with nil value.
+func (n *nfsVolume) CollectCreateEvents() (string, error) {
+
+	volumeData, err := n.getVolumeData()
+	if err != nil {
+		return "", err
+	}
+
+	if !n.isSupportedDataType() {
+		return "", errors.Errorf("data type %q is not supported. Supported types %v", n.dataType, supportedDataTypes)
+	}
+
+	if volumeData.NFSPVC != nil {
+		volumeData.NFSPVC.DeletionTimestamp = nil
+		volumeData.NFSPVC.DeletionGracePeriodSeconds = nil
+	}
+
+	volumeData.NFSPV.DeletionTimestamp = nil
+	volumeData.NFSPV.DeletionGracePeriodSeconds = nil
+
+	volumeData.BackingPVC.DeletionTimestamp = nil
+	volumeData.BackingPVC.DeletionGracePeriodSeconds = nil
+
+	volumeData.BackingPV.DeletionTimestamp = nil
+	volumeData.BackingPV.DeletionGracePeriodSeconds = nil
+
+	createData := &NFSCreateVolumeData{
+		VolumeProvisioned: volumeData,
+	}
+	rawData, err := json.Marshal(createData)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to marshal create volume events")
+	}
+	return string(rawData), nil
+}
+
+func (n *nfsVolume) CollectDeleteEvents() (string, error) {
+	// TODO: Is below check required?
```

Comment:
we can drop this check since the caller is taking care of this.

Reply:
Removed check